k4FWCore/components/EventHeaderCreator.cpp

```diff
@@ -26,25 +26,13 @@ EventHeaderCreator::EventHeaderCreator(const std::string& name, ISvcLocator* svc
                   "Name of the EventHeaderCollection that will be stored in the output root file.");
 }
 
-StatusCode EventHeaderCreator::initialize() {
-  if (Gaudi::Algorithm::initialize().isFailure())
-    return StatusCode::FAILURE;
-  return StatusCode::SUCCESS;
-}
-
-StatusCode EventHeaderCreator::execute(const EventContext&) const {
-  static int eventNumber = 0;
+StatusCode EventHeaderCreator::execute(const EventContext& ctx) const {
+  auto eventNumber = ctx.evt();
```

Comment:
Who sets the event number in this case? And to what values under what circumstances? Is this deterministic and reproducible?

Reply:
It is ordinal event number created by the `EventProcessor` (event loop).  Both `HiveSlimEventLoopMgr` and `EventLoopMgr` keep track of the number of events run so far, update the number in `nextEvent` and put it in `EventContext`

Being deterministic and reproducible depends on the scheduler. 
- With event context solution that would mean there is no another `Reader` alg executed (from different event processed in parallel) between `EventProcessor::nextEvent()` and `Reader` algorithm for a given event.
- With  the atomic counter solution that would mean there is no `EventHeaderCreator` alg executed between `Reader` alg and `EventHeaderCreator` alg for a given event

Reply:
With "there is no", do you mean "there must be no"? Can this be guaranteed?

Reply:
Sorry, I meant there should be no overlap between those operations in order to have no mismatch between ordinal number of event in an input file and number used in header.

For example if there are just three events in an input file and the `EventHeaderCreator` is used to create the headers for them , then if there is an overlap the between those operations then the event numbers in headers could be any permutation of (0,1,2) with no reproducibility (but at least there will be no duplicates as in the current `EventHeaderCreator`)

 I don't think these can be guaranteed with `AvalancheSchedulerSvc` (beside limiting the thread pool and concurrency slots)

Reply:
There was a discussion on the key4hep meeting a week ago that it should be acceptable to not have deterministic event ids when running in a multithreaded environment